src/collapse/perf.rs

```diff
@@ -154,29 +188,96 @@ impl Collapse for Folder {
+            line.clear();
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+            if line.starts_with('#') {
+                continue;
+            }
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else if self.in_event {
+                self.on_stack_line(line)
+            } else {
+                self.on_event_line(line)
+            }
         }
+        Ok(())
     }
-}
 
-impl Folder {
-    fn on_line(&mut self, line: &str) {
-        if !self.in_event {
-            self.on_event_line(line)
-        } else {
-            self.on_stack_line(line)
-        }
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        debug_assert!(self.occurrences.is_concurrent());
+        debug_assert!(self.nthreads > 1);
+
+        let mut buf = Vec::with_capacity(CAPACITY_INPUT_BUFFER);
+        reader.read_to_end(&mut buf)?;
+
+        let mut input = Input::new(
+            buf,
+            self.nthreads,
+            Self::identify_stack_locations(&mut self.event_filter_state),
+        )?;
+
+        crossbeam::thread::scope(|scope| {
+            let mut handles = Vec::with_capacity(input.nthreads());
+            let (sender, receiver) = crossbeam::channel::bounded(input.nthreads());
+            for chunk in input.chunks() {
+                let event_filter_state = self.event_filter_state.clone();
+                let occurrences = self.occurrences.clone();
+
+                let annotate_jit = self.annotate_jit;
+                let annotate_kernel = self.annotate_kernel;
+                let include_addrs = self.include_addrs;
+                let include_pid = self.include_pid;
+                let include_tid = self.include_tid;
+
+                let sender = sender.clone();
+
+                let handle = scope.spawn(move |_| {
+                    let mut folder = Folder {
+                        // state
+                        cache_line: Vec::default(),
+                        event_filter_state,
+                        in_event: false,
+                        occurrences,
+                        skip_stack: false,
+                        stack: VecDeque::default(),
+                        pname: String::new(),
+
+                        // options
+                        annotate_jit,
+                        annotate_kernel,
+                        include_addrs,
+                        include_pid,
+                        include_tid,
+                        nthreads: 1,
+                    };
+                    let result = folder.collapse_single_threaded(chunk);
+                    sender.send(result).unwrap();
+                });
+                handles.push(handle);
+            }
+            for handle in handles {
+                receiver.recv().unwrap()?;
+                handle.join().unwrap();
```

Comment:
Come to think of it, doesn't crossbeam just directly give you back the return value of the spawned closure? You may not even need the channel (here or in dtrace).

Reply:
It's the error I'm trying to recover, actually. I know that if a thread with crossbeam threads panics, you get back the text of the error it panicked on, but not the actual error type. I can play around with this to see if we can just get what we want from whatever crossbeam recovers from the thread panicking.

Reply:
So crossbeam's `ScopedJoinHandle`'s `join` method returns a `Result<T, Box<dyn Any + 'static + Send>>`. You get the error variant if the thread panicked. There are ways to try to downcast the `Any` you get back to a `&str` or `String` in order to recover the text of the panic, but isn't it just easier to do what I'm doing, i.e. create a channel for what `collapse` returns so we send a nice `io::Result<()>` back to the main thread and then deal with the error, if any, there?